src/hydrovu.rs

```diff
@@ -0,0 +1,408 @@
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Names {
+    parameters: HashMap<String, String>,
+    units: HashMap<String, String>,
+}
+
+// Location is documented at https://www.hydrovu.com/public-api/docs/index.html
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Location {
+    description: String,
+    id: i64,
+    name: String,
+    gps: LatLong,
+}
+
+// LatLong is documented at https://www.hydrovu.com/public-api/docs/index.html
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct LatLong {
+    latitude: f64,
+    longitude: f64,
+}
+
+// LocationReadings is a batch of timeseries.
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct LocationReadings {
+    location_id: i64,
+    parameters: Vec<ParameterInfo>,
+}
+
+// ParameterInfo is a single timeseries.
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct ParameterInfo {
+    custom_parameter: bool,
+    parameter_id: String,
+    unit_id: String,
+    readings: Vec<Reading>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Reading {
+    timestamp: i64,
+    value: f64,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Vu {
+    pub units: HashMap<String, String>,
+    pub params: HashMap<String, String>,
+    pub locations: Vec<Location>,
+}
+
+fn location_fields() -> Vec<FieldRef> {
+    vec![
+        Arc::new(Field::new("description", DataType::Utf8, false)),
+        Arc::new(Field::new("id", DataType::UInt64, false)),
+        Arc::new(Field::new("name", DataType::Utf8, false)),
+        Arc::new(Field::new(
+            "gps",
+            DataType::Struct(Fields::from(vec![
+                Field::new("latitude", DataType::Float64, false),
+                Field::new("longitude", DataType::Float64, false),
+            ])),
+            false,
+        )),
+    ]
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct Mapping {
+    index: String,
+    value: String,
+}
+
+fn mapping_fields() -> Vec<FieldRef> {
+    vec![
+        Arc::new(Field::new("index", DataType::Utf8, false)),
+        Arc::new(Field::new("value", DataType::Utf8, false)),
+    ]
+}
+
+fn evar(name: &str) -> Result<String, Box<dyn Error>> {
+    Ok(env::var(name).with_context(|| format!("{name} is not set"))?)
+}
+
+impl Client {
+    fn fetch_json<T: for<'de> serde::Deserialize<'de>>(&self, url: String) -> ClientCall<T> {
+        ClientCall::<T> {
+            client: self,
+            url: url,
+            next: Some("".to_string()),
+            phan: PhantomData,
+        }
+    }
+}
+
+const HYDROVU_CLIENT_ID_ENV: &str = "HYDROVU_CLIENT_ID";
+const HYDROVU_CLIENT_SECRET_ENV: &str = "HYDROVU_CLIENT_SECRET";
+
+fn creds() -> Result<(String, String), Box<dyn Error>> {
+    Ok((
+        evar(HYDROVU_CLIENT_ID_ENV)?,
+        evar(HYDROVU_CLIENT_SECRET_ENV)?,
+    ))
+}
+
+pub fn new_client() -> Result<Client, Box<dyn Error>> {
```

Comment:
Generally `Box<dyn Error>` is frowned upon because it requires an allocation and and introduces a vtable. Typically you'd use a so-called "error crate" or the approach I personally prefer, define a top level error type and implement `From<ForeignError>` for that new top level error type.

In pseudo rust:
```rust
use arrow::ArrowError;
pub enum MyCrateError {
    Arrrow(ArrowError),
    Timeout,
    IO(std::io::Error)
}

impl From<ArrowError> for MyCrateError {
    fn from(value: ArrowError) -> Self {
        Self::Arrow(value)
    }
}
```

Reply:
Thank you. I admit I didn't like typing "Box<dyn Error>" very much and this approach you outlined sounds much better. Will study!

Reply:
Another option is to use the `thiserror` crate for your crates/libraries and `anyhow` for an application.